File: construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    Dictionary<int,int> dict = new Dictionary<int,int>();
    int preindex=0;
    public TreeNode BuildTree(int[] preorder, int[] inorder) {
        int idx =0;
        foreach(int i in inorder)
        {
            dict.Add(i,idx++);
        }
        
        return helper(0,inorder.Length,preorder,inorder);
        
    }
    
    public TreeNode helper(int left,int right,int[] preorder,int[] inorder)
    {
        if(left==right)
        {
            return null;
        }
        
        int root_val = preorder[preindex];
        TreeNode root = new TreeNode(root_val);
        
        int index = dict[root_val];
        
        preindex++;
        root.left  = helper(left,index,preorder,inorder);
        root.right = helper(index+1,right,preorder,inorder);
        return root;
    }
}

File: src/ConstructBinaryTreefromPreorderandInorderTraversal.java

/*
 * The complexity of the algorithm is O(n) where n is no of elements.
 * Space complexity is O(n) where n is no of elements.
 */

import java.util.HashMap;

public class ConstructBinaryTreefromPreorderandInorderTraversal {

	public class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;

		TreeNode() {
		}

		TreeNode(int val) {
			this.val = val;
		}

		TreeNode(int val, TreeNode left, TreeNode right) {
			this.val = val;
			this.left = left;
			this.right = right;
		}
	}

	HashMap<Integer, Integer> map;
	int idx;

	public TreeNode buildTree(int[] preorder, int[] inorder) {

		if (preorder.length == 0)
			return null;

		map = new HashMap<>();

		for (int i = 0; i < inorder.length; i++) {
			map.put(inorder[i], i);
		}

		return helper(preorder, inorder, 0, inorder.length - 1);
	}

	private TreeNode helper(int[] preorder, int[] inorder, int start, int end) {

		if (start > end)
			return null;

		int rootval = preorder[idx];
		idx++;

		TreeNode root = new TreeNode(rootval);

		int rIdx = map.get(rootval);

		root.left = helper(preorder, inorder, start, rIdx - 1);
		root.right = helper(preorder, inorder, rIdx + 1, end);

		return root;

	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub

		ConstructBinaryTreefromPreorderandInorderTraversal objIn = new ConstructBinaryTreefromPreorderandInorderTraversal();

	}

}
